### PR TITLE
fix Issue 22721 - importC: some gnu builtins are rejected

### PR DESCRIPTION
**Pull request by WalterBright**
Not a whole lot we can do with the ones other than bswap.

### CODE REVIEW

File: src/__builtins.di [hunk 1]
```diff
@@ -68,4 +68,14 @@ version (DigitalMars)
...
     alias __builtin_fabsl = core.stdc.math.fabsl;
+
+    import core.bitop;
+
+    alias __builtin_bswap32 = core.bitop.bswap;
+    alias __builtin_bswap64 = core.bitop.bswap;
+
+    // Stub these out to no-ops
+    enum  __builtin_constant_p = 0; // should be something like __traits(compiles, enum X = expr)
+    long  __builtin_expect(long exp, long c) { return 0; }
```
Comment (thewilsonator) [inline src/__builtins.di:79]: `return 0;` is not what `__builtin_expect` does. It should `return exp;`

File: src/__builtins.di [hunk 2]
```diff
@@ -68,4 +68,14 @@ version (DigitalMars)
...
     alias __builtin_fabsf = core.stdc.math.fabsf;
     alias __builtin_fabsl = core.stdc.math.fabsl;
+
+    import core.bitop;
+
+    alias __builtin_bswap32 = core.bitop.bswap;
+    alias __builtin_bswap64 = core.bitop.bswap;
+
+    // Stub these out to no-ops
+    enum  __builtin_constant_p = 0; // should be something like __traits(compiles, enum X = expr)
```
Comment (thewilsonator) [inline src/__builtins.di:78]: signature is wrong should be `int   __builtin_constant_p( int value) { return 0; }` (with a pessimism that no value is know at compile time).

**Review by thewilsonator — CHANGES_REQUESTED** [review]

**Comment by WalterBright** [comment]
https://issues.dlang.org/show_bug.cgi?id=22721

https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html

File: src/__builtins.di (hunk 1 above)
Comment (WalterBright) [reply src/__builtins.di:79]: yah, you're right. I should go to bed now :-)

**Comment by thewilsonator** [comment]
> https://issues.dlang.org/show_bug.cgi?id=22721
>
> https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html

...and? 

You still have to get the form of their usage correct.

**Review by thewilsonator — APPROVED** [review]

File: src/__builtins.di (hunk 2 above)
Comment (WalterBright) [reply src/__builtins.di]: actually it should be a template, as the argument can be of any type

Comment (ibuclaw) [reply src/__builtins.di]: Do we support `enum foo = (int) => 0;`?

File: src/__builtins.di [hunk 3]
```diff
@@ -68,4 +68,14 @@ version (DigitalMars)
     alias __builtin_fabs  = core.stdc.math.fabs;
     alias __builtin_fabsf = core.stdc.math.fabsf;
     alias __builtin_fabsl = core.stdc.math.fabsl;
+
+    import core.bitop;
+
+    alias __builtin_bswap32 = core.bitop.bswap;
+    alias __builtin_bswap64 = core.bitop.bswap;
```
Comment (ssvb) [inline src/__builtins.di:75]: Mapping both `__builtin_bswap32` and `__builtin_bswap64` to the same function doesn't look safe. Consider the following C code:

```C
#include <assert.h>

int main()
{
    int x = 123;
    long long bs32 = __builtin_bswap32(x);
    long long bs64 = __builtin_bswap64(x);
    assert(bs32 == 0x7b000000ll);
    assert(bs64 == 0x7b00000000000000ll);
    return 0;
}
```

And D implementation of this feature probably needs unit tests.

Comment (ibuclaw) [reply src/__builtins.di:75]: They're D overloads.

Comment (ssvb) [reply src/__builtins.di:75]: > They're D overloads.

Yes, they are. But this doesn't make them safe. In my C example above, the lines `long long bs32 = __builtin_bswap32(x);` and `long long bs64 = __builtin_bswap64(x);`have exactly the same argument type (int) and exactly the same return value type (long long). If you alias them to the same `bswap` function, then they will naturally return exactly the same result for the same input. But the result must be different, as confirmed by the assert lines. So in order to work correctly, the implementation of `__builtin_bswap64` needs to explicitly cast its argument to a 64-bit type.

Additionally please consider the following D code:
```D
import std.stdio, std.conv, core.bitop;

void main()
{
    int x = 123;

    writefln("%x", x.to!short.bswap);    // prints 7b000000
    writefln("%x", x.to!ushort.bswap);   // prints 7b000000

    writefln("%x", x.to!int.bswap);      // prints 7b000000
    writefln("%x", x.to!uint.bswap);     // prints 7b000000

    writefln("%x", bswap(cast(long)x));  // prints 7b000000
    writefln("%x", bswap(cast(ulong)x)); // prints 7b00000000000000

    writefln("%x", x.to!long.bswap);     // prints 7b00000000000000
    writefln("%x", x.to!ulong.bswap);    // prints 7b00000000000000
}
```
The printed results for `.to!short`, `.to!ushort` and `cast(long)` seem to be fishy. But this probably needs to be filed as a separate bug.

**Review by ssvb — CHANGES_REQUESTED** [review]
So my suggestion is to use something like this instead of the alias (also only 2 lines of code):
```D
uint  __builtin_bswap32(uint value)  { return core.bitop.bswap(value); }
ulong __builtin_bswap64(ulong value) { return core.bitop.bswap(value); }
```

File: src/__builtins.di (hunk 3 above)
Comment (WalterBright) [reply src/__builtins.di:75]: It's not to the same function, but to the same template. Importing templates works in ImportC.

**Comment by WalterBright** [comment]
> my suggestion is

The `alias` does the same thing, with the advantage of it not needing to link to a function.

**Comment by ssvb** [comment]
> The alias does the same thing, with the advantage of it not needing to link to a function.

@WalterBright No, it doesn't do the same thing. I'm not allowed to register in bugzilla with a gmail address, so pasting my bugreport here: `ImportC: __builtin_bswap64 is implemented incorrectly`

foobarmod.c
```C
long long foobar()
{
  return __builtin_bswap64(123);
}
```

testc.c
```C
#include <stdio.h>

long long foobar();

int main()
{
  printf("%llx\n", foobar());
  return 0;
}
```

testd.d
```D
import std.stdio, foobarmod;

void main()
{
  writefln("%x", foobar());
}
```

Compile and run (on a 64-bit linux system):
```
$ ./dmd testd.d foobarmod.c && ./testd 
7b000000

$ gcc testc.c foobarmod.c && ./a.out
7b00000000000000
```
The results differ and this isn't good.

**Comment by ibuclaw** [comment]
Isn't that a bug in your code though? You are using `%x` to format a 64bit value... (also printf is the worst form of verification)

**Comment by ssvb** [comment]
> Isn't that a bug in your code though?

I don't think so.

> You are using %x to format a 64bit value... 

Using %x to format integers of any size is supported in D language, see https://dlang.org/library/std/format/write.html

> (also printf is the worst form of verification)

Well, my first example used `assert`. If you are not very comfortable with `assert` and `writefln/printf`, then you can suggest something else. But to be honest, I would prefer somebody else to take a look and see whether my examples and explanations about what's wrong with this pull request make any sense to them.

**Comment by adamdruppe** [comment]
On Thu, Feb 03, 2022 at 12:40:11PM -0800, Siarhei Siamashka wrote:
> Using %x to format integers of any size is supported in D language, see https://dlang.org/library/std/format/write.html

writef and printf are completely separate functions.

**Comment by ssvb** [comment]
Yes, they are. But @ibuclaw is complaining about the use of `writefln` in D code, so my documentation link is relevant.

**Comment by WalterBright** [comment]
@ssvb you're right. `__builtin_bswap64(123)` winds up calling bswap32 because `123` is an int, not a long long.

Do you want the honor of doing a pull request to fix it?

**Comment by ssvb** [comment]
@WalterBright Thanks. But if it's possible, I would prefer to remain an observer for a little bit longer and watch how it's done properly. So it would be best if somebody else makes a pull request. I also wonder, do these new builtin functions need (or benefit from) any attributes, such as `pure`?

**Comment by WalterBright** [comment]
https://issues.dlang.org/show_bug.cgi?id=22735

**Pull request merged**